rts/motoko-rts/src/gc/mark_compact/bitmap.rs

```diff
@@ -2,55 +2,126 @@ use crate::mem_utils::memzero;
 
+/* How the Wasm-heap maps to the bitmap
+
+  +---- RTS stack ----+---- Motoko statics ----+---- Dynamic heap --~~--+ Heap limit
+                  (prefix words)                   bitmap lives here -> | BM |
+                                                                       /      \
+                                                                      /        \
+                                       each bit represents a word -> /...bits...\
+                                          in the dynamic heap
+
+## Marking with absolute addresses
+
+When marking, we pass an absolute pointer (i.e. address space relative), for speed.
+Internally the bitmap is kept in a (non-moving) blob behind the dynamic heap (DH).
+To efficiently mark the right bit in the bitmap, we maintain a pointer that points
+_before the start of the bitmap_ such that using the `/%8`-operation on the DH
+absolute word number will address the right bit:
+
+
+       +---- BITMAP_FORBIDDEN_PTR         +---- BITMAP_PTR
+       v                                  v
+       ,          (forbidden)             ,...................... bitmap ..........~~....,
+       |   heap_prefix_words / 8 bytes    |               heap_size / 32 bytes           |
+       |   get_bitmap_forbidden_size()    |                   BITMAP_SIZE                |
+       ^                                  ^                                              ^
+      /           ^^^^^^^^^^^              \                                             |
+     /            corresponds               \                 ^^^^^^^^^^^               /
+    /                                        \                corresponds              /
+   /                                          \                                       /
+  +---- Rts stack ----+---- Motoko statics ----+--------- Dynamic heap ----------~~--+
+                                               !
+                                               ! 32-byte aligned
+
+
+Debug assertions guard the forbidden bytes from access, as this area potentially
+physically overlaps with the Motoko dynamic heap.
+
+## The alignment caveat
+
+For this scheme to work, it is essential that the start of the DH is an address that
+is divisible by 32 (implying that `heap_prefix_words % 8 == 0`). Otherwise the
+`/%8`-operation on the DH's starting address will not yield the least significant bit
+in the BM, and thus the sweep operation will be off.
+
+## Example calculation
+
+Assume the DH is at 0x80000. Assume heap limit being at 0xB0000. Then the BM thus
+could be placed at 0xB0004. Since the heap_prefix_words is 0x20000,
+BITMAP_FORBIDDEN_PTR = 0xB0004 - 0x20000 / 8 = 0xAC004.
+
+Now let's mark the address 0x80548 in the DH. Its absolute word number is 0x20152.
+The `(0x20152 / 8, 0x20152 % 8)`-rule gives a bit position 2 with byte offset 0x402A,
+thus we mark bit 2 in byte 0xAC004 + 0x402A = 0xB002E, which is physically in the BM.
```

Comment:
```suggestion
thus we mark bit 2 in byte 0xAC004 + 0x402A = 0xB002E, which is in the BM.
```

Same as above. Maybe it's me being obsessive, but when I read a qualifier like "physically" here, I think about what else it could be. In this case, I think about what does being "virtually in the BM" mean (opposite of "virtually in the BM"). I think this distinction doesn't make sense, so the qualifier is confusing to me.

As usual, feel free to ignore.

Reply:
I think this is done.